### PR TITLE
Fixes #2985 by logging a specific message in case of timeout and using an async logger.

### LINKED ISSUE #2985: Marathon cannot schedule task when log writing is blocked
I don't think these two should be related that much. We saw this when running Marathon (both 0.11.1 and 0.14.0RC3) under docker that is very busy, see docker/docker#19209.

```
2016-01-14T00:21:02.000Z 36s54 [2016-01-14 00:16:47,422] ERROR error while matching '44bc368b-40c7-4669-8a25-340efc507a2e-O513897' (mesosphere.marathon.core.launcher.impl.OfferProcessorImpl:59)
2016-01-14T00:21:02.000Z 36s54 at akka.actor.Scheduler$$anon$7.run(Scheduler.scala:117)
2016-01-14T00:21:02.000Z 36s54 at akka.pattern.PromiseActorRef$$anonfun$1.apply$mcV$sp(AskSupport.scala:335)
2016-01-14T00:21:02.000Z 36s54 akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://marathon/user/offerMatcherManager#168703490]] after [1000 ms]
2016-01-14T00:21:02.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.run(Scheduler.scala:375)
2016-01-14T00:21:02.000Z 36s54 at java.lang.Thread.run(Thread.java:745)
2016-01-14T00:21:02.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
2016-01-14T00:21:02.000Z 36s54 at scala.concurrent.BatchingExecutor$class.execute(BatchingExecutor.scala:109)
2016-01-14T00:21:02.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:599)
2016-01-14T00:21:02.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
2016-01-14T00:21:03.000Z 36s54 at java.lang.Thread.run(Thread.java:745)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.nextTick(Scheduler.scala:423)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.executeBucket$1(Scheduler.scala:419)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(Scheduler.scala:467)
2016-01-14T00:21:03.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
2016-01-14T00:21:03.000Z 36s54 [2016-01-14 00:16:47,422] ERROR error while matching '44bc368b-40c7-4669-8a25-340efc507a2e-O513893' (mesosphere.marathon.core.launcher.impl.OfferProcessorImpl:59)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.run(Scheduler.scala:375)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.Scheduler$$anon$7.run(Scheduler.scala:117)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.executeBucket$1(Scheduler.scala:419)
2016-01-14T00:21:03.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:599)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(Scheduler.scala:467)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.nextTick(Scheduler.scala:423)
2016-01-14T00:21:03.000Z 36s54 akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://marathon/user/offerMatcherManager#168703490]] after [1000 ms]
2016-01-14T00:21:03.000Z 36s54 [2016-01-14 00:16:47,422] ERROR error while matching '44bc368b-40c7-4669-8a25-340efc507a2e-O513891' (mesosphere.marathon.core.launcher.impl.OfferProcessorImpl:59)
2016-01-14T00:21:03.000Z 36s54 akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://marathon/user/offerMatcherManager#168703490]] after [999 ms]
2016-01-14T00:21:03.000Z 36s54 at akka.pattern.PromiseActorRef$$anonfun$1.apply$mcV$sp(AskSupport.scala:335)
2016-01-14T00:21:03.000Z 36s54 at akka.pattern.PromiseActorRef$$anonfun$1.apply$mcV$sp(AskSupport.scala:335)
2016-01-14T00:21:03.000Z 36s54 at scala.concurrent.BatchingExecutor$class.execute(BatchingExecutor.scala:109)
2016-01-14T00:21:03.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:599)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.Scheduler$$anon$7.run(Scheduler.scala:117)
2016-01-14T00:21:03.000Z 36s54 at java.lang.Thread.run(Thread.java:745)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(Scheduler.scala:467)
2016-01-14T00:21:03.000Z 36s54 at scala.concurrent.BatchingExecutor$class.execute(BatchingExecutor.scala:109)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.executeBucket$1(Scheduler.scala:419)
2016-01-14T00:21:03.000Z 36s54 at akka.actor.LightArrayRevolverScheduler$$anon$8.nextTick(Scheduler.scala:423)
2016-01-14T00:21:03.000Z 36s54 at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
2016-01-14T00:21:03.000Z 36s54 at java.lang.Thread.run(Thread.java:745)
```

The minimum thing that should be done is error message. It should probably point to the possible source of the error and give some clues how to improve. Ex:

```
Taking too long [1000ms] to match offers, see --max_whatever.
```

Here `--max_whatever` can have some explanations in docs.

### PR DESCRIPTION
**Pull request by aquamatthias**

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala [hunk 1]
```diff
@@ -57,7 +58,10 @@ private[launcher] class OfferProcessorImpl(
       .recover {
         case NonFatal(e) =>
           matchErrorsMeter.mark()
-          log.error(s"error while matching '${offer.getId.getValue}'", e)
+          val message =
+            if (e.isInstanceOf[AskTimeoutException]) "in time. Extend with --max_offer_matching_timeout."
+            else s"${e.getClass.getSimpleName}: ${e.getMessage}"
+          log.warn(s"Could not process offer '${offer.getId.getValue}' $message")
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala]: Shouldn't other error still be logged as errors?

Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala]: And can we maybe generalize this code to apply it elsewhere where we ask with a timeout?

File: src/main/resources/logback.xml [hunk 2]
```diff
@@ -6,7 +6,11 @@
             <pattern>[%date] %level %message \(%logger:%thread\)%n</pattern>
         </encoder>
     </appender>
+    <appender name="async" class="ch.qos.logback.classic.AsyncAppender">
```
Comment (kolloch) [inline src/main/resources/logback.xml]: Do we flush this on termination?

http://logback.qos.ch/manual/appenders.html#AsyncAppender
http://logback.qos.ch/manual/configuration.html#stopContext

**Comment by kolloch** [comment]
Some comments.

File: src/main/resources/logback.xml (hunk 2 above)
Comment (aquamatthias) [inline src/main/resources/logback.xml]: Added a logback shutdown hook.

File: src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala (hunk 1 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala]: Handle Timeout and other exceptions different.

**Comment by kolloch** [comment]
Thanks! :rocket:

**Pull request merged**